import 'dart:io';

import 'package:get/get.dart';
import 'package:jumper/src/Core/constants/handle_error_constant.dart';
import 'package:jumper/src/Core/resources/data_state.dart';
import 'package:jumper/src/Core/services/network/network_exceptions.dart';
import 'package:jumper/src/Features/shared/NotificationsFeature/bloc/model/notificationsModel.dart';
import 'package:jumper/src/Features/shared/NotificationsFeature/bloc/services/notifications_api_services.dart';

class NotificationsRepo {
  final NotificationsApiServices _notificationsApiServices =
      Get.put(NotificationsApiServices());

  Future<DataState<List<NotificationsModel>>>? fetchNotifications() async {
    try {
      final httpResponse = await _notificationsApiServices.fetchNotifications();
      if (((httpResponse.statusCode == HttpStatus.ok) ||
              (httpResponse.statusCode == HttpStatus.created) ||
              (httpResponse.statusCode == HttpStatus.accepted)) &&
          (httpResponse.data['status'] ?? false)) {
        if (httpResponse.data['data'] != null) {
          List<NotificationsModel> notifications = [];
          for (var item in httpResponse.data['data']) {
            notifications.add(NotificationsModel.fromJson(item));
          }
          return DataSuccess<List<NotificationsModel>>(notifications);
        } else {
          return DataFailed(ErrorModel(
              errorTitle: httpResponse.data['message'] ?? '',
              errorType: ErrorType.dataEmpty));
        }
      }
      return DataFailed(ErrorModel(
          errorTitle: httpResponse.data['message'] ?? '',
          errorType: ErrorType.unKnown));
    } on NetworkDisconnectException {
      return DataFailed(ErrorModel(
          errorTitle: networkDisconnectException,
          errorType: ErrorType.networkConnection));
    } on ServerSideException {
      return DataFailed(ErrorModel(
          errorTitle: serverSideException, errorType: ErrorType.serverSide));
    } on UnKnownException {
      return DataFailed(ErrorModel(
          errorTitle: unKnownException, errorType: ErrorType.unKnown));
    } catch (e) {
      return DataFailed(
          ErrorModel(errorTitle: '$e', errorType: ErrorType.unKnown));
    }
  }
}
